### PR TITLE
Parallel extract merge

### PR DESCRIPTION
**Pull request by macintux**
Incorporate changes to support 2i parallel extract and return_body.

### CODE REVIEW

File: include/riak_core_vnode.hrl [hunk 1]
```diff
@@ -35,3 +35,16 @@
 -define(FOLD_REQ, #riak_core_fold_req_v2).
 
 -type handoff_dest() :: {riak_core_handoff_manager:ho_type(), {partition(), node()}}.
+
+%% An integer, and the number of bits to shift it left to treat it as
+%% a mask in the 2^160 key space
```
Comment (jadeallenx) [inline include/riak_core_vnode.hrl:40]: Should probably point out that much of these data structures have a verbose explanation in the `riak_core_coverage_plan` module

File: src/riak_core_coverage_fsm.erl [hunk 2]
```diff
@@ -205,6 +205,21 @@ maybe_start_timeout_timer(Timeout) ->
...
+                     PVC,
+                     ReqId,
+                     Service,
+                     Request);
+find_plan(VNodeTarget, NVal, PVC, ReqId, Service, _Request) ->
+    riak_core_coverage_plan:create_plan(VNodeTarget,
+                                        NVal,
+                                        PVC,
+                                        ReqId,
+                                        Service).
```
Comment (jadeallenx) [inline src/riak_core_coverage_fsm.erl:223]: :+1: Like this

File: src/riak_core_coverage_plan.erl [hunk 3]
```diff
@@ -110,120 +293,433 @@ create_plan(VNodeSelector, NVal, PVC, ReqId, Service) ->
...
+
+%% @private
+%% Take our replacement traditional coverage chunks and consolidate it
+%% into a traditional coverage plan (that another layer will rechunk,
+%% but whatchagonnado)
+dechunk_traditional_replacement(Coverage) ->
+    {
+      lists:map(fun({{Vnode, Node}, _Filters}) ->
+                        {Vnode, Node}
+                end, Coverage),
```
Comment (jadeallenx) [inline src/riak_core_coverage_plan.erl:366]: Torn between asking you to replace this with a list comprehension and leaving as-is.

File: src/riak_core_coverage_plan.erl [hunk 4]
```diff
@@ -110,120 +293,433 @@ create_plan(VNodeSelector, NVal, PVC, ReqId, Service) ->
...
+    [].
+
+%% @private
+create_subpartition_plan(VNodeTarget, NVal, Count, PVC, ReqId, Service, CHBin, AvailNodeFun) ->
+    MaskBSL = data_bits(Count),
+
+    %% Calculate an offset based on the request id to offer the
+    %% possibility of different sets of VNodes being used even when
+    %% all nodes are available.
+    Offset = ReqId rem NVal,
```
Comment (jadeallenx) [inline src/riak_core_coverage_plan.erl:511]: Should this use a better source of randomness? Not maybe cryptographic strength randomness, but the request term is going to be fairly monotonic and not very random.

File: src/riak_core_coverage_plan.erl [hunk 5]
```diff
@@ -110,120 +293,433 @@ create_plan(VNodeSelector, NVal, PVC, ReqId, Service) ->
...
+map_pvc_vnodes([{VnodeIdx, Node}|Tail], SubpID, PVC, Accum) ->
+    map_pvc_vnodes(Tail, SubpID, PVC-1,
+                   [{ VnodeIdx, Node, SubpID }] ++ Accum).
+
+
+%% @private
+%% Adding an offset while keeping the result inside [0, Top). Adding
+%% Top to the left of `rem' allows offset to be negative without
+%% violating the lower bound of zero
+add_offset(Position, Offset, Top) ->
```
Comment (jadeallenx) [inline src/riak_core_coverage_plan.erl:583]: Love this comment. :candy:

**Comment by jadeallenx** [comment]
Tests and dialyzer run clean. Nice work.

File: src/riak_core_coverage_plan.erl (hunk 4 above)
Comment (macintux) [inline src/riak_core_coverage_plan.erl:511]: This is how we do coverage plans today, and given the small space (typically only 3 choices with n_val=3) I'm not worried about better randomness

File: src/riak_core_coverage_plan.erl (hunk 3 above)
Comment (macintux) [inline src/riak_core_coverage_plan.erl:366]: Deadlines, gonna have to say "no" even though it's a valid thought

File: include/riak_core_vnode.hrl (hunk 1 above)
Comment (macintux) [inline include/riak_core_vnode.hrl:40]: Added

**Comment by jadeallenx** [comment]
:+1: 3183449

**Comment by macintux** [comment]
@borshop merge

**Comment by macintux** [comment]
@borshop merge

**Pull request merged**